enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,177 @@
+to get a download link for `odo` or `oc`.  The in-cluster cli-manager will take this a step further to handle installing binaries on the user's system, 
+in the $HOME directory, that are known compatible with a cluster's version.  The challenge is that we need a mechanism that works in disconnected environments.
+Also, this has created a burden of maintenance for the console team.  Furthermore, we want to offer a mechanism that other CLI authors can leverage to deliver tools in-cluster.  
+CLI-manager will replace the current consoleclidownloads mechanism and will extend its function.
+
+2.  Central repository where all artifacts are stored - [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+currently is where crc, oc, ocp-dev-preview, ocp, odo, and serverless artifacts are published.  This is currently how OpenShift CLIs are published.  The ConsoleCLIDownloads
+CRs reference this index.  The challenge is offering these in disconnected environments and automating the download and extraction of the artifacts.  
+
+3.  Images for each CLI. The CLI in-cluster manager would provide an extract mechanism for each CLI image, similar to how we currently 
+`oc adm release extract --command oc` and `oc image extract`.   A Custom Resource for each CLI would provide the in-cluster CLI manager information about each CLI and its image.
+In disconnected, the images will be available through a mirrored local registry.  The logic for `oc image extract` can be re-used and extended for extracting the CLIs. 
+    * CLI owners build images to package artifacts similar to https://github.com/sallyom/clustercli/tree/master/hack/cli-images
+    * CLIs would be managed by users through extending the oc commands we currently have to extract `oc` and `openshift-install` binaries from the release payload.
+    This would provide the function we need, but looking to the future there is an upstream effort we can use to provide an in-cluster CLI manager while also providing 
+    a mechanism that the Kubernetes community as a whole can utilize.  That brings us to option 4.
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
```

Comment:
krew is familiar to the `kubectl` users however some of the CLIs we provide are for developer who are not familiar with `kubectl`. Also krew is a `kubectl` plugin manager. I am not sure how well it applies to providing CLIs. 

I am not sure if `krew`, or trying to provide a CLI manager on the client's local machine should be a primary concern for this.

Reply:
getting back to this.
I've updated this proposal to indicate the initial implementation will be for each CLI author to 
1) build CLI image to package artifacts
2) provide a ClusterCLI Custom Resource describing the CLI and it's image/location
3) register and manage the CR via an operator

`krew` may in the future facilitate the install and/or provide the index  but not for inital implementation.